### PR TITLE
refactor: refactor the logic for building android projects 

### PR DESCRIPTION
**Pull request by Fatme**
## PR Checklist

- [ ] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [ ] You have signed the [CLA](http://www.nativescript.org/cla).
- [ ] All existing tests are passing: https://github.com/NativeScript/nativescript-cli/blob/master/CONTRIBUTING.md#contribute-to-the-code-base
- [ ] Tests for the changes are included.

## What is the new behavior?

Fixes/Implements/Closes #[Issue Number].

### CODE REVIEW

**Comment by Fatme** [comment]
test cli-smoke

File: lib/definitions/gradle.d.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+interface IGradleBuildService {
+	buildProject(projectRoot: string, buildConfig: IAndroidBuildConfig): Promise<void>;
+	cleanProject(projectRoot: string, buildConfig: IAndroidBuildConfig): Promise<void>;
+}
+
+interface IGradleBuildArgsService {
+	getBuildTaskArgs(buildConfig: IAndroidBuildConfig): string[];
+	getCleanTaskArgs(buildConfig: IAndroidBuildConfig): string[];
+}
```
Comment (KristianDD) [inline lib/definitions/gradle.d.ts]: New line

File: lib/services/android-project-service.ts [hunk 2]
```diff
@@ -277,76 +232,13 @@ export class AndroidProjectService extends projectServiceBaseLib.PlatformProject
...
-			buildOptions.push(`-Palias=${settings.keyStoreAlias}`);
-			buildOptions.push(`-Ppassword=${settings.keyStoreAliasPassword}`);
-			buildOptions.push(`-PksPassword=${settings.keyStorePassword}`);
-		}
-
-		return buildOptions;
-	}
-
-	private getBuildOptions(configurationFilePath?: string): Array<string> {
-		this.$androidToolsInfo.validateInfo({ showWarningsAsErrors: true, validateTargetSdk: true });
```
Comment (KristianDD) [inline lib/services/android-project-service.ts]: Is this validation obsolete?

Comment (Fatme) [reply lib/services/android-project-service.ts]: Yes, it is executed before the execution of each command, so no need to validate it here again.

File: lib/services/android-project-service.ts [hunk 3]
```diff
@@ -522,81 +404,14 @@ export class AndroidProjectService extends projectServiceBaseLib.PlatformProject
...
-			return true;
-		}
-
-		const androidStudioCompatibleTemplate = "3.4.0";
-		const normalizedPlatformVersion = `${semver.major(platformVersion)}.${semver.minor(platformVersion)}.0`;
-
-		return semver.gte(normalizedPlatformVersion, androidStudioCompatibleTemplate);
-	}
-
 	private getLegacyAppResourcesDestinationDirPath(projectData: IProjectData): string {
```
Comment (KristianDD) [inline lib/services/android-project-service.ts]: Maybe we can remove the support for legacy App_Resources structure for 6.0

File: lib/services/android/gradle-build-args-service.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		args.unshift("clean");
+
+		return args;
+	}
+
+	private getBaseTaskArgs(buildConfig: IAndroidBuildConfig): string[] {
+		const args = this.getBuildLoggingArgs();
+
+		const toolsInfo = this.$androidToolsInfo.getToolsInfo();
+		args.push(
```
Comment (KristianDD) [inline lib/services/android/gradle-build-args-service.ts]: Are we omitting `-PtargetSdk` on purpose?

File: lib/services/android/gradle-build-service.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	constructor(
+		private $childProcess: IChildProcess,
+		private $gradleBuildArgsService: IGradleBuildArgsService,
+		private $gradleCommandService: IGradleCommandService,
+	) { super(); }
+
+	public async buildProject(projectRoot: string, buildConfig: IAndroidBuildConfig): Promise<void> {
+		const buildTaskArgs = this.$gradleBuildArgsService.getBuildTaskArgs(buildConfig);
+		const spawnOptions = { emitOptions: { eventName: constants.BUILD_OUTPUT_EVENT_NAME }, throwError: true };
+		const gradleCommandOptions = { cwd: projectRoot, message: "Gradle build...", stdio: buildConfig.buildOutputStdio, spawnOptions };
```
Comment (KristianDD) [inline lib/services/android/gradle-build-service.ts]: Previously there was a default value for the stdio - inherit. Does it still default to that down the execution line?

File: lib/services/android-project-service.ts (hunk 3 above)
Comment (Fatme) [reply lib/services/android-project-service.ts]: We discussed this and decided to keep the support for legacy structure for `6.0.0` release

File: lib/services/android/gradle-build-service.ts (hunk 5 above)
Comment (Fatme) [reply lib/services/android/gradle-build-service.ts]: We still have a default value - https://github.com/NativeScript/nativescript-cli/pull/4537/files#diff-235c99870cec07a0b3afcf34af51be12R10

Comment (KristianDD) [reply lib/services/android/gradle-build-service.ts]: Saw it in executeCommand :)

File: lib/services/android/gradle-command-service.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+		const childProcessOptions = { cwd, stdio };
+		const gradleExecutable = this.$hostInfo.isWindows ? "gradlew.bat" : "./gradlew";
+
+		const result = await this.executeCommandSafe(gradleExecutable, gradleArgs, childProcessOptions, spawnOptions);
+
+		return result;
+	}
+
+	private async executeCommandSafe(gradleExecutable: string, gradleArgs: string[], childProcessOptions: any, spawnOptions: any) {
```
Comment (KristianDD) [inline lib/services/android/gradle-command-service.ts]: spawnFromEvent: ISpawnFromEventOptions

File: lib/services/android/gradle-build-args-service.ts (hunk 4 above)
Comment (Fatme) [reply lib/services/android/gradle-build-args-service.ts]: We discussed the support for `targetSdk` version and as there are some unclear things around it we decided to do it in another PR.

**Pull request merged**